abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,1996 @@
+
+* params:
+  - `i32 (size_t *) return_num_elements`
+  - `i32 (size_t *) return_buffer_size`
+* returns:
+  - `i32 (`[`wasi_errno_t`]`) errno`
+
+Returns `return_num_elements` of environment variables, and
+`return_buffer_size` sufficient to serialize them and decorators.
+
+Returned `errno` value is:
+- `SUCCESS` on success.
+- `FAULT` when `return_num_elements` and/or `return_buffer_size`
+  point to invalid memory address.
+
+
+#### `wasi_snapshot_preview1.environ_get`
+
+* params:
+  - `i32 (uint8_t **) return_array`
+  - `i32 (uint8_t *) return_buffer`
+* returns:
+  - `i32 (`[`wasi_errno_t`]`) errno`
+
+Returns serialized environment variables.
+
+Returned `errno` value is:
+- `SUCCESS` on success.
+- `FAULT` when `return_array` and/or `return_buffer` point to
+  invalid memory address.
+
+
+## Buffers
+
+Access to buffers using functions in this section is restricted to
+specific callbacks:
+
+- `HTTP_REQUEST_BODY` can be read and modified in
+  [`proxy_on_request_body`] (or for as long as request processing
+  is paused from it).
+- `HTTP_RESPONSE_BODY` can be read and modified in
+  [`proxy_on_response_body`] (or for as long as response processing
+  is paused from it).
+- `DOWNSTREAM_DATA` can be read and modified in
+  [`proxy_on_downstream_data`] (or for as long as upstream processing
+  is paused from it).
+- `UPSTREAM_DATA` can be read and modified in
+  [`proxy_on_downstream_data`] (or for as long as downstream processing
+  is paused from it).
+- `HTTP_CALL_RESPONSE_BODY` can be read in
+  [`proxy_on_http_call_response`].
+- `GRPC_CALL_MESSAGE` can be read in [`proxy_on_grpc_receive`].
+- `VM_CONFIGURATION` can be read in [`proxy_on_vm_start`].
+- `PLUGIN_CONFIGURATION` can be read in [`proxy_on_configure`].
+- `FOREIGN_FUNCTION_ARGUMENTS` can be read in
+  [`proxy_on_foreign_function`].
+
+
+### Functions exposed by the host
+
+#### `proxy_set_buffer_bytes`
+
+* params:
+  - `i32 (`[`proxy_buffer_type_t`]`) buffer_id`
+  - `i32 (uint32_t) offset`
+  - `i32 (uint32_t) size`
+  - `i32 (const uint8_t *) value_data`
+  - `i32 (size_t) value_size`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Sets content of the buffer `buffer_id` to the provided value
+(`value_data`, `value_size`) replacing `size` bytes starting
+at `offset` in the existing buffer.
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for unknown `buffer_id`.
+- `NOT_FOUND` when requested `buffer_id` isn't available.
+- `INVALID_MEMORY_ACCESS` when `value_data` and/or `value_size`
+  point to invalid memory address.
+
+
+#### `proxy_get_buffer_bytes`
+
+* params:
+  - `i32 (`[`proxy_buffer_type_t`]`) buffer_id`
+  - `i32 (uint32_t) offset`
+  - `i32 (uint32_t) max_size`
+  - `i32 (uint8_t **) return_value_data`
+  - `i32 (size_t *) return_value_size`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Retrieves up to` max_size` bytes starting at `offset` from the buffer
+`buffer_id`.
+
+Returned `status` value is:
+- `OK` on success.
```

Comment:
Should we doc the behavior when the return buffer is empty? (For example, request without a body)